pkgs/development/haskell-modules/make-package-set.nix

```diff
@@ -176,6 +176,17 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
 
     callHackage = name: version: callPackageKeepDeriver (self.hackage2nix name version);
 
+    # This function does not depend on all-cabal-hashes and therefore will work
+    # for any version that has been released on hackage as opposed to only
+    # versions released before whatever version of all-cabal-hashes you happen
+    # to be currently using.
+    callHackageDirect = {pkg, ver, sha256}@args:
+      let pkgver = "${pkg}-${ver}";
+      in self.callCabal2nix pkg (pkgs.fetchzip {
```

Comment:
Why does above use `callPackageKeepDeriver` instead of `callCabal2nix`? Should this do the same?

Reply:
When I change it to the following:

```
    callHackageDirect = {pkg, ver, sha256}@args:
      let pkgver = "${pkg}-${ver}";
      in callPackageKeepDeriver (pkgs.fetchzip {
           url = "http://hackage.haskell.org/package/${pkgver}/${pkgver}.tar.gz";
           inherit sha256;
         }) {};
```

I get this error:

```
error: opening file '/nix/store/3vlj5ajyc56hishmz7qna2n8mmpr24h1-source/default.nix': No such file or directory
```

Reply:
I think you have to do `callPackageKeepDeriver` on top of the call to cabal2nix. Anyway that can be saved for a future improvement.

Reply:
I'm not sure why `callHackage` uses `callPackageKeepDeriver`, but I played around with it and couldn't get it to work using `callPackageKeepDeriver` even trying @shlevy's suggestion of calling it on top of `cabal2nix`.  If you look at `callCabal2nix`, you'll set that it ultimately calls `callPackageKeepDeriver`, so I think this formulation should be fine.